plugins/chat/app/services/chat/update_user_last_read.rb

```diff
@@ -4,76 +4,76 @@ module Chat
     step :publish_new_last_read_to_clients
 
     # @!visibility private
     class Contract
       attribute :message_id, :integer
-      attribute :user_id, :integer
       attribute :channel_id, :integer
 
-      validates :message_id, :user_id, :channel_id, presence: true
+      validates :message_id, :channel_id, presence: true
     end
 
     private
 
-    def fetch_active_membership(user_id:, channel_id:, **)
-      Chat::UserChatChannelMembership.includes(:user, :chat_channel).find_by(
-        user_id: user_id,
-        chat_channel_id: channel_id,
-        following: true,
-      )
+    def fetch_channel(contract:, **)
+      ::Chat::Channel.find_by(id: contract.channel_id)
+    end
+
+    def fetch_active_membership(guardian:, channel:, **)
+      ::Chat::ChannelMembershipManager.new(channel).find_for_user(guardian.user, following: true)
     end
 
-    def invalid_access(guardian:, membership:, **)
-      guardian.can_join_chat_channel?(membership.chat_channel)
+    def invalid_access(guardian:, active_membership:, **)
+      guardian.can_join_chat_channel?(active_membership.chat_channel)
     end
 
-    def ensure_message_id_recency(message_id:, membership:, **)
-      !membership.last_read_message_id || message_id >= membership.last_read_message_id
+    def ensure_message_exists(channel:, contract:, **)
+      ::Chat::Message.with_deleted.exists?(chat_channel_id: channel.id, id: contract.message_id)
     end
 
-    def ensure_message_exists(channel_id:, message_id:, **)
-      Chat::Message.with_deleted.exists?(chat_channel_id: channel_id, id: message_id)
+    def ensure_message_id_recency(contract:, active_membership:, **)
+      !active_membership.last_read_message_id ||
+        contract.message_id >= active_membership.last_read_message_id
     end
 
-    def update_last_read_message_id(message_id:, membership:, **)
-      membership.update!(last_read_message_id: message_id)
+    def update_last_read_message_id(contract:, active_membership:, **)
+      active_membership.update!(last_read_message_id: contract.message_id)
     end
 
-    def mark_associated_mentions_as_read(membership:, message_id:, **)
-      Notification
+    def mark_associated_mentions_as_read(active_membership:, contract:, **)
+      ::Notification
         .where(notification_type: Notification.types[:chat_mention])
-        .where(user: membership.user)
+        .where(user: active_membership.user)
         .where(read: false)
         .joins("INNER JOIN chat_mentions ON chat_mentions.notification_id = notifications.id")
         .joins("INNER JOIN chat_messages ON chat_mentions.chat_message_id = chat_messages.id")
-        .where("chat_messages.id <= ?", message_id)
-        .where("chat_messages.chat_channel_id = ?", membership.chat_channel.id)
+        .where("chat_messages.id <= ?", contract.message_id)
+        .where("chat_messages.chat_channel_id = ?", active_membership.chat_channel.id)
         .update_all(read: true)
```

Comment:
Should we extract this as an action (it's very similar to the one in mark_all_user_channels...)? Fine to be a in a different PR

Reply:
True, will extract in this same PR :+1: